### ISSUE TITLE
The run function in the task protocol has too many responsibilities

### ISSUE THREAD

**Issue by Ivan-Johnson** [Ivan-Johnson/zam#18]
The `run` function in the `task` protocol of `scheduler.py` is responsible for both running the class and computing the time at which it should next be run. These responsibilities should be split into separate `getNextActionionableDatetime` (?) and `run` functions.

Besides just making for a cleaner interface, this would also allow the scheduler to fully initialize its prioritized heap without having to actually run each task.

**Comment by Ivan-Johnson** [comment]
When there is just a single `run` function, crude period tasks can be implemented by simply returning a constant. Splitting `run` into two functions complicates this as you will have to track the time that it was last run at. This common use case can be implemented using another instance of the `task` protocol.

    class PeriodicTask(task):
        last_run : Typing.optional[datetime.datetime]
        period : datetime.timedelta

        def __init__(self, period:datetime.timedelta):
            self.last_run = None
            self.period = period
    
        def getNextActionionableDatetime() -> datetime.datetime:
            if self.last_run is None:
                return now()
            else:
                return last_run + period
        
        def run(self) -> None:
            self.last_run = now()

This `PeriodicTask` would best be extended using composition rather than inheritance. There is a fairly persuasive argument to be made that this composition should be done by having `PeriodicTask` take a function pointer as input which it would then run in the `run` function, but I'm inclined against that approach because it introduces the need for a couple additional parameters to `PeriodicTask`:

1. Should the task run immediately, or wait for one period first?
2. Should the `last_run` date be updated before or after the function pointer is called?

Come to think of it though, the above pseudocode will drift due to inaccuracies in the scheduler and competing tasks running when this one was scheduled to. This already necessecitates a parameter to indicate whether we want this drift or not. The additional complexity of implementing and testing that parameter makes me inclined to think that it actually isn't worth having a `PeriodicTask` at all; tasks can just implement that behavior themselves in about the same number of lines of code, and honestly it would probably be more readable that way anyways.

TL;DR: We could add a common `PeriodicTask` class to save a couple of lines of code from each periodic task, but it probably isn't worth the effort

**Comment by Ivan-Johnson** [comment]
Related, `PrioritizedTask` should be renamed to `__PrioritizedTask`
